File: java/completed/euler0023.java
import lib.EulerTime;
import lib.EulerUtil;

import java.util.ArrayList;

/**
 * Created by Holden on 12/8/2016. - COMPLETED
 *
 * Problem 23 - Non-Abundant Sums
 *
 */

public class Prob0023 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        int answer = nonAbundantSums(28123);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static int nonAbundantSums(int bound) {
        ArrayList<Integer> abundantSums = new ArrayList<>();
        for (int i = 2; i <= bound; i++) {
            ArrayList<Integer> divisors = EulerUtil.properDivisors(i);
            int sum = 0;
            for (int divisor: divisors) {
                sum+= divisor;
            }

            if (sum > i) {
                abundantSums.add(i);
            }
        }

        int[] nonAbundantSums = new int[bound+1];
        for (int i = 0; i <= bound; i++) {
            nonAbundantSums[i] = i;
        }

        for (int i: abundantSums) {
            for (int j: abundantSums) {
                if (i + j <= bound) {
                    nonAbundantSums[i+j] = 0;
                }
            }
        }

        int nonAbundantSum = 0;
        for (int i: nonAbundantSums) {
            nonAbundantSum += i;
        }

        return nonAbundantSum;
    }
}

File: python/completed/euler0023.py
###
#
# Created by Holden on 12/18/2015
#
# 
#
# Problem 23 - Non-Abundant Sums
#
###


def proper_divisors(n):
    divisors = []

    for i in range(1, n // 2 + 1):
        if n % i == 0:
            divisors.append(i)

    return divisors

    
def non_abundant_sums(bound):
    abundant_sums = []

    for i in range(2, bound):
        divisors = proper_divisors(i)
        divisor_sum = 0
        for divisor in divisors:
            divisor_sum += divisor

        if divisor_sum > i:
            abundant_sums.append(i)

    final_sums = [i for i in range(bound+1)]

    for i in abundant_sums:
        for j in abundant_sums:
            if i + j <= bound:
                final_sums[i+j] = 0

    return sum(final_sums)


print(non_abundant_sums(28123))
